Training24/intlist.c

```diff
@@ -0,0 +1,200 @@
+// ------------------------------------------------------------------------------------------------
+// Training ~ A training program for new joiners at Metamation, Batch - July 2024.
+// Copyright (c) Metamation India.
+// ------------------------------------------------------------------
+// intlist.c
+// Program on Linked list functions.
+// ------------------------------------------------------------------------------------------------
+
+#include <malloc.h>
+#include "intlist.h"
+
+// Global variable to store the last error code
+ int mlastError = SUCCESS;
+
+// Creating a new list
+// Allocate memory for a new list structure and initialize the list as empty
+IntList* Create () {
+   IntList* newList = (IntList*)malloc (sizeof (IntList));
+   if (newList == NULL) {
+      mlastError = MEMORY_ALLOCATION_FAILED;
+      return NULL;
+   }
+   newList->head = NULL; ; // Intialilze head pointer as null
+   mlastError = SUCCESS;
+   return newList;
+}
+
+// Delete the entire list and free memory
+void Delete (IntList* list) {
+   if (list == NULL) {
+      mlastError = NULL_POINTER;
+      return;
+   }
+   Node* current = list->head;
+   Node* next;
+   while (current != NULL) {  // Traverse the list and free each node
+      next = current->next;
+      free (current);
+      current = next;
+   }
+   free (list);
+   mlastError = SUCCESS;
+}
+
+// Add a new element to the end of the list
+void Add (IntList* list, int data) {
+   if (list == NULL) {
+      mlastError = NULL_POINTER;
+      return;
+   }
+   Node* newNode = (Node*)malloc (sizeof (Node));
+   if (newNode == NULL) {
+      mlastError = MEMORY_ALLOCATION_FAILED;
+      return;
+   }
+   newNode->data = data;
+   newNode->next = NULL;
+   if (list->head == NULL) {  //If the list is empty, set the new node as the head
+      list->head = newNode;
+   }
+   else {
+      Node* current = list->head;
+      while (current->next != NULL) {  // Traverse to the end of the list
+         current = current->next;
+      }
+      current->next = newNode;
+   }
+   mlastError = SUCCESS;
+}
+
+// Insert a new element at a specified index (zero-based)
+void Insert (IntList* list, int index, int data) {
+   if (list == NULL) {
+      mlastError = NULL_POINTER;
+      return;
+   }
+   if (index < 0 || index > Count (list)) {

```

Comment:
This is good. See [guard](https://en.wikipedia.org/wiki/Guard_(computer_science))